Allow numbers.Integral instead of only built-in int for QuantumRegister index

arunraja-hub opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Fixes #3929

### Details and comments
-I changed line 120/121 of register.py such that it includes numbers.Integral as an allowable type for an index of a QuantumRegister. This is my first issue! Excited to contribute more
Issue: https://github.com/Qiskit/qiskit-terra/issues/3929

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=4591) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=4591) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=4591) it.</sub>

Cryoris commented:
Thanks for the contribution @arunraja-hub! Can you add a test that checks passing different types of integers as register sizes? I.e. a test that fails without your PR but works afterwards.

arunraja-hub commented:
Hi @Cryoris , Thank you! Could you provide more info on implementing the test? is there is any documentation on that?

Cryoris commented on qiskit/circuit/register.py:
```diff
@@ -117,7 +118,7 @@ def __getitem__(self, key):
             QiskitIndexError: if the `key` is not in the range `(0, self.size)`.
         """
-        if not isinstance(key, (int, slice, list)):
+        if not isinstance(key, (int, slice, list, numbers.Integral)):
```
`numbers.Integral` covers `int` already, we don't need both 
```suggestion
        if not isinstance(key, (numbers.Integral, slice, list)):
```

Cryoris commented on releasenotes/notes/allowIntegerforQuantumRegisterSize-2e9a6cc033bc962e.yaml:
```diff
@@ -0,0 +1,13 @@
+---
+prelude: >
+    allow use of type integer for for QuantumRegister size 
+
+issues:
+  - |
+    QuantumRegister allows only int, slice, list to denote size
+
```
The prelude is only added upon release, we don't need to add this in a PR. 🙂  Also you don't need to refer to the issue, I think this section is for issues with the release and does not refer to GitHub issues.
```suggestion
```

Cryoris commented on releasenotes/notes/allowIntegerforQuantumRegisterSize-2e9a6cc033bc962e.yaml:
```diff
@@ -0,0 +1,13 @@
+
+fixes:
+  - |
+    Added numbers.Integral as another allowable type to denote size of QuantumRegister
```
You can use markdown formatting here 👍 
```suggestion
    Use numbers.Integral instead of Python's built-in ``int`` as 
    input type to denote the size of a ``QuantumRegister``.
```

Cryoris commented:
I don't think there's a dedicated documentation for the tests, but essentially a test is a short snippet of code that checks your new feature works. In a bugfix test, you ideally have a snippet that didn't work before but now does. E.g. something like
```python
def test_register_int_types(self):
    ints = [int(2), int64(2), np.int(2), np.integer(2), ... ] # all int types you can think of
    for size in ints:
        with self.subTest(size=size):  # technically not necessary, but nicer for debugging if you add this
            qr = QuantumRegister(size)
            cr = ClassicalRegister(size)
            self.assertEqual(qr.size, 2)
            self.assertEqual(cr.size, 2)
```
See [test/python/circuit/test_circuit_registers.py](https://github.com/Qiskit/qiskit-terra/blob/master/test/python/circuit/test_circuit_registers.py) for where you could implement the test and how others look like.

Let me know if you have any questions!

arunraja-hub commented:
`test_register_int_types` passed when I ran the command `make test`
However, according to github's checks above there seem to be 2 cases that are failing. Why is this the case?

Cryoris commented:
The CI also runs a linter and stylechecker on the code, you can do this locally by doing `make lint` and `make style` in the command line. The output is
```
test/python/circuit/test_circuit_registers.py:61:1: W293 blank line contains whitespace
test/python/circuit/test_circuit_registers.py:63:77: W291 trailing whitespace
test/python/circuit/test_circuit_registers.py:65:44: E202 whitespace before ']'
test/python/circuit/test_circuit_registers.py:65:46: W291 trailing whitespace
test/python/circuit/test_circuit_registers.py:67:42: W291 trailing whitespace
```
If you fix this, the tests will passs.

Cryoris commented on test/python/circuit/test_circuit_registers.py:
```diff
@@ -59,6 +59,17 @@ def test_qarg_numpy_int_size(self):
+    def test_register_int_types(self):
+        """Test attempt to pass different types of integer as register size
+        """
+        ints = [int(2), float(2), np.int(2)]
```
```suggestion
        ints = [int(2), float(2), np.int(2), np.int32(2), np.int64(2)]
```

arunraja-hub commented:
@Cryoris Completed the tests for int32 and int64. I believe this PR is all set

Cryoris commented:
Actually I misunderstood the goal, I though the _size_ should accept all types not the _index_. The tests should be changed accordingly to check that all these integer types work as indices of `QuantumRegister` and `ClassicalRegister`. You could e.g. change the test to do something like
```python
for index in [np.int(2), np.int32(2), ... ]:
        # ... sub test as before, and the same for classical register
        qr = QuantumRegister(3)
        self.assertEqual(qr[index], qr[2])
```

Cryoris commented on releasenotes/notes/allowIntegerforQuantumRegisterSize-2e9a6cc033bc962e.yaml:
```diff
@@ -0,0 +1,6 @@
+fixes:
+  - |
+    Use numbers.Integral instead of Python's built-in ``int`` as 
+    input type to denote the size of a ``QuantumRegister``.
```
```suggestion
    supported index to ``QuantumRegister`` and ``ClassicalRegister``.
```

arunraja-hub commented:
I changed the test as follows:
```
    def test_register_int_types(self):
        """Test attempt to pass different types of integer as indices of QuantumRegister and ClassicalRegister
        """
        ints = [int(2), float(2), np.int(2), np.int32(2), np.int64(2)]
        for index in ints:
            with self.subTest(index=index):
                qr = QuantumRegister(4)
                cr = ClassicalRegister(4)
                self.assertEqual(qr[index], qr[2])
                self.assertEqual(cr[index], cr[2])
```
But I get the following error after running `make test`:
```
Traceback (most recent call last):
  File "C:\Users\User\Documents\qiskit\os\qiskit-terra\test\python\circuit\test_circuit_registers.py", line 70, in test_register_int_types
    self.assertEqual(qr[index], qr[2])
  File "C:\Users\User\Documents\qiskit\os\qiskit-terra\qiskit\circuit\register.py", line 122, in __getitem__
    raise CircuitError("expected integer or slice index into register")
qiskit.circuit.exceptions.CircuitError: 'expected integer or slice index into register'
```

Cryoris commented:
Good, that's what tests are for -- to get error messages like this 🙂  In this case it is most likely the `float` argument. If you want you could add a more meaningful error message here, saying that floats cannot be passed as arguments. To also capture complex numbers and numpy floats you could use `numbers.Number` which captures all python numbers. See also [here](https://docs.python.org/3.8/library/numbers.html) for more info.

Cryoris commented on qiskit/circuit/register.py:
```diff
@@ -117,7 +118,7 @@ def __getitem__(self, key):
             QiskitIndexError: if the `key` is not in the range `(0, self.size)`.
         """
-        if not isinstance(key, (int, slice, list)):
+        if not isinstance(key, (numbers.Number, int, numbers.Integral, slice, list)):
```
It might be a little confusing to allow floats as index. Could we change this to allow only integers and raise an error otherwise? What I meant before was something like 
```suggestion
        if not isinstance(key, (numbers.Integral, slice, list)):
            if isinstance(key, numbers.Number):
                raise CircuitError("register index cannot be a float/complex, must be integer or slice")
```
this way we have a clearer error message than before if the user passed in a float.

arunraja-hub commented:
@Cryoris  The lint checks are failing here but I am getting a 10/10 when I run make lint

kdk commented:
> @Cryoris The lint checks are failing here but I am getting a 10/10 when I run make lint

Here's the error from CI:
https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=16515&view=logs&j=7684b0c9-af38-5400-184f-6880bd8a4dd7&t=9d95188b-4e59-58d8-f7c4-3f1fbd4209ad&l=11

arunraja-hub commented:
Ths issue was that I had a newline. So I removed. Now the lint error is that I have removed a newline. So it is contradictory
https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=16523&view=logs&j=7684b0c9-af38-5400-184f-6880bd8a4dd7&t=9d95188b-4e59-58d8-f7c4-3f1fbd4209ad&l=12

kdk commented on qiskit/circuit/register.py:
```diff
@@ -152,4 +153,4 @@ def __eq__(self, other):
     def __hash__(self):
         """Make object hashable, based on the name and size to hash."""
-        return self._hash
+        return self._hash
```
```suggestion
        return self._hash

```
The linter is warning about a missing newline character at the end of the file. I think this should fix it.

Cryoris commented on qiskit/circuit/register.py:
```diff
@@ -117,7 +118,7 @@ def __getitem__(self, key):
             QiskitIndexError: if the `key` is not in the range `(0, self.size)`.
         """
-        if not isinstance(key, (int, slice, list)):
+        if not isinstance(key, (numbers.Integral, int, slice, list)):
```
```suggestion
        if not isinstance(key, (numbers.Integral, slice, list)):
```

arunraja-hub commented:
Hi @Cryoris, May I confirm that this PR is complete?  Also, could you assign this issue to me? Thank you

Cryoris commented:
Yes, it is! Thanks a lot for the contribution 🚀

arunraja-hub commented:
Thank you @Cryoris 
Are there other issues I can work on next?
